### ISSUE TITLE
Enhancement: Support CodePath rule listener methods

### ISSUE THREAD

**Issue by NotWoods** [typescript-eslint/typescript-eslint#6993]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Rules created using `createRule` currently don't have TypeScript support for the [Code Path Analysis](https://eslint.org/docs/latest/extend/code-path-analysis) events. This would be great to add since they still work fine, its just that TypeScript complains.

I'd like to be able to use `onCodePathStart`, `onCodePathEnd`, `onCodePathSegmentStart`, `onCodePathSegmentEnd`, and `onCodePathSegmentLoop` as well as use an interface for CodePath and CodePathSegment,

### Fail

```typescript
N/A
```

### Pass

```typescript
createRule({
  meta: { ... },
  defaultOptions: [],
  create(context) {
    return {
      /**
       * @param {import('eslint').Rule.CodePath} codePath
       * @param {import('@typescript-eslint/utils').TSESTree.Node} node
       */
      onCodePathStart(codePath, node) {
        const isEffect = ASTUtils.isFunction(node) && isUseEffectCall(node.parent);
        functionInfoStack.push({ codePath, node, isEffect, listeners: [] });
      }
    }
  }
})
```

**Comment by bradzacher** [comment]
The reason we never added these is two fold:

1) it's really, really rare that people want to use them - I don't think I've seen more than half a dozen rules that actually use them (including the ones in ESLint core).

2) their form (`(codePath: { id: string }, node: TSESTree.Node) => void`) is incompatible with the node selector signature (`(node: TSESTree.Node) => void`). This causes TS errors because the code path signatures cannot be assigned to the index signature.

See existing Pr: https://github.com/typescript-eslint/typescript-eslint/pull/2930

Because the usecase is so very rare and it's a difficult problem to solve - we've just opted to not solve it yet.

If you can think of a way to add the definitions **without harming the DevX of node selectors** then I'm more than happy to have them added to the types!

I must really stress that it's important that we do not impact the DevX of selectors - there are thousands of node selectors written with our types - so we mustn't make them less safe, nor must we make them harder to use just to add types for this fraction of a fraction of a percent case.

**Comment by Zamiell** [comment]
Just chiming in here to say that code paths are pretty nice to have.

Code paths are used in a few official ESLint rules, namely [`no-fallthrough`](https://github.com/eslint/eslint/blob/main/lib/rules/no-fallthrough.js). As you can see from looking at the source code for the rule, it is incredibly short, because all of the heavy-lifting is being done by the code paths feature. In other words, the code paths feature allows us to trivially determine when a particular switch statement case arm is a "fallthrough" case arm.

This functionality is also used in other rules that have to do with switch statements. For example, look at the source code for the [`newline-between-switch-case`](https://github.com/lukeapage/eslint-plugin-switch-case/blob/master/docs/rules/newline-between-switch-case.md) rule from [`eslint-plugin-switch-case`](https://github.com/lukeapage/eslint-plugin-switch-case). It also uses code paths to trivially detect the fallthrough case.

Of course, code paths are not just useful for rules about switch statements, but for other situations too. They are nice because we can detect certain kinds of situations without having the rule to be type aware, which obviously greatly hinders a rule's performance impact.

As a first baby step, I think that this project should export a `CodePath` interface at the very least. That would allow plugin authors to perform some manual type assertion. For example:

```ts
import type { TSESTree } from "@typescript-eslint/utils";

export const myRule = createRule<Options, MessageIds>({
  create(context) {
    return {
      // We must type assert `never` to `CodePath` because using the nodes inside of it can cause problems with TypeScript.
      onCodePathStart(codePath as TSESTree.CodePath) {}
    };
  },
```

**Comment by bradzacher** [comment]
To clarify - it's not that we don't think it's useful - there are some rare usecase for them and they can help solve some problems or simplify others!

Instead what were saying is that unfortunately TS isn't designed to support this usecase well - you can't define an index signature that has one value but also define members with a different, incompatible value.

So in order for us to add the signatures we'd have to break all the existing code that relies on the index signature - which is obviously an intractable solution - breaking 99.99% of cases to support 0.01% isn't good!

I'd welcome someone adding the codepath types - but we just can't add the signatures.

**Comment by armano2** [comment]
few months ago i tried to add it, but i got stuck with exactly same issue, see https://github.com/typescript-eslint/typescript-eslint/pull/6679

**Comment by Zamiell** [comment]
@armano2 What about a PR for just the interfaces?

**Comment by bradzacher** [comment]
> [I'd welcome someone adding the codepath types - but we just can't add the signatures.](https://github.com/typescript-eslint/typescript-eslint/issues/6993#issuecomment-1694524075)

**Comment by Zamiell** [comment]
@JoshuaKGoldberg I don't think this issue should be closed actually, since I only added the interfaces in my PR

**Comment by bradzacher** [comment]
We can't add the listener methods without big changes in how TS works - we rely on the `[string]: (TSESTree.Node) => void` signature to provide support for esquery selectors and that signature is fundamentally incompatible with the code path selector function signatures (none of which take `TSESTree.Node` as their first argument).

So unless TS adds a way for us to support both without degrading the experience of esquery selectors - we will have to close this as "fixed as best as we can for now".
